File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

enum ItemType {
  Normal,
  // Conjured,
  BackstagePasses,
  FineCheeses,
  Legendaries,
}
class ItemCategory {
  item: Item;
  type: ItemType;

  constructor(item: Item) {
    this.item = item;
    this.type = ItemCategory.getItemType(item);
  }

  static CHEESES = ['aged brie']
  static LEGENDARIES = ['sulfuras, hand of ragnaros']

  static getItemType(item: Item): ItemType {
    // if item starts with Conjured
    if (item.name.toLowerCase().startsWith('backstage passes')){
      return ItemType.BackstagePasses;
    }

    // if item name is a fine cheese
    if (this.CHEESES.includes(item.name.toLowerCase())){
      return ItemType.FineCheeses;
    }

    if (this.LEGENDARIES.includes(item.name.toLowerCase())){
      return ItemType.Legendaries;
    }
    return ItemType.Normal;
  }

  updateQuality() {
    switch(this.type) {
      case ItemType.BackstagePasses:
        return this.updateBackstageQuality();
      case ItemType.Normal:
        return this.updateNormalQuality();
      case ItemType.FineCheeses:
        return this.updateCheeseQuality();
      case ItemType.Legendaries:
        return this.updateLegendaryQuality();
    }
  }

  updateNormalQuality() {
    const quality = this.item.quality;
    if (this.item.sellIn > 0) {
      this.item.quality = Math.max(quality - 1, 0);
    } else {
      this.item.quality = Math.max(quality - 2, 0);
    }

    this.item.sellIn -= 1;
  }

  updateCheeseQuality() {
    const quality = this.item.quality;
    this.item.quality = Math.min(quality + 1, 50);
  }

  updateBackstageQuality() {
    const quality = this.item.quality;
    if(this.item.sellIn <= 0) {
      this.item.quality = 0;  
    }
    else if (this.item.sellIn < 11) {
      this.item.quality = Math.min(quality + 2, 50);
    } else if (this.item.sellIn < 6) {
      this.item.quality = Math.min(quality + 3, 50);
    } else {
      this.item.quality = Math.min(quality + 1, 50);
    }

    this.item.sellIn -= 1;
  }

  updateLegendaryQuality() {}
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      const category = new ItemCategory(this.items[i]);
      category.updateQuality();

      // if (this.items[i].name != 'Aged Brie' && this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
      //   if (this.items[i].quality > 0) {
      //     if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
      //       this.items[i].quality = this.items[i].quality - 1
      //     }
      //   }
      // } else {
      //   if (this.items[i].quality < 50) {
      //     this.items[i].quality = this.items[i].quality + 1
      //     if (this.items[i].name == 'Backstage passes to a TAFKAL80ETC concert') {
      //       if (this.items[i].sellIn < 11) {
      //         if (this.items[i].quality < 50) {
      //           this.items[i].quality = this.items[i].quality + 1
      //         }
      //       }
      //       if (this.items[i].sellIn < 6) {
      //         if (this.items[i].quality < 50) {
      //           this.items[i].quality = this.items[i].quality + 1
      //         }
      //       }
      //     }
      //   }
      // }
      // if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
      //   this.items[i].sellIn = this.items[i].sellIn - 1;
      // }
      // if (this.items[i].sellIn < 0) {
      //   if (this.items[i].name != 'Aged Brie') {
      //     if (this.items[i].name != 'Backstage passes to a TAFKAL80ETC concert') {
      //       if (this.items[i].quality > 0) {
      //         if (this.items[i].name != 'Sulfuras, Hand of Ragnaros') {
      //           this.items[i].quality = this.items[i].quality - 1
      //         }
      //       }
      //     } else {
      //       this.items[i].quality = this.items[i].quality - this.items[i].quality
      //     }
      //   } else {
      //     if (this.items[i].quality < 50) {
      //       this.items[i].quality = this.items[i].quality + 1
      //     }
      //   }
      // }
    }

    return this.items;
  }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items
        self.legendary_items = ["Sulfuras, Hand of Ragnaros"]
        self.increases_over_time_items = ["Aged Brie", "Backstage passes to a TAFKAL80ETC concert"]

    def update_quality(self):
        for item in self.items:
            name = item.name
            if self.does_increase_over_time(name) and item.quality <= 50:
                if name == "Aged Brie":
                    item.increase_brie_quality()
                if name == "Backstage passes to a TAFKAL80ETC concert":
                    item.increase_backstage_pass_quality()
            elif item.quality > 0 and not self.is_legendary(name):
                if item.conjured:
                item.quality = item.quality - 1
            if not self.is_legendary(name):
                item.sell_in = item.sell_in - 1
            if item.sell_in < 0:
                if name != "Aged Brie":
                    if name != "Backstage passes to a TAFKAL80ETC concert":
                        if item.quality > 0 and not self.is_legendary(name):
                            item.quality = item.quality - 1
                    else:
                        item.quality = item.quality - item.quality
                elif item.quality < 50:
                    item.quality = item.quality + 1

    def is_legendary(self, item_name):
        if item_name in self.legendary_items:
            return True
        else:
            return False

    def does_increase_over_time(self, item_name):
        if item_name in self.increases_over_time_items:
            return True
        else:
            return False




class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)

    def increase_backstage_pass_quality(self):
        if self.sell_in <= 5:
            self.quality += 3
        elif self.sell_in <= 10:
            self.quality += 2
        else:
            self.quality += 1
        if self.quality > 50:
            self.quality = 50

    def increase_brie_quality(self):
        self.quality += 1
        if self.quality > 50:
            self.quality = 50
